compiler/noirc_evaluator/src/ssa/opt/fill_internal_slices.rs

```diff
@@ -0,0 +1,705 @@
+
+        let instruction_id = instruction;
+        let (instruction, call_stack) = self.inserter.map_instruction(instruction_id);
+        let new_array_op_instr = match instruction {
+            Instruction::ArrayGet { index, .. } => {
+                Instruction::ArrayGet { array: new_array, index }
+            }
+            Instruction::ArraySet { index, value, .. } => {
+                Instruction::ArraySet { array: new_array, index, value }
+            }
+            _ => panic!("Expected array set"),
+        };
+
+        (new_array_op_instr, call_stack)
+    }
+
+    fn attach_slice_dummies(
+        &mut self,
+        typ: &Type,
+        value: Option<ValueId>,
+        is_parent_slice: bool,
+        max_sizes: &[usize],
+    ) -> ValueId {
+        match typ {
+            Type::Numeric(_) => {
+                if let Some(value) = value {
+                    self.inserter.resolve(value)
+                } else {
+                    let zero = FieldElement::zero();
+                    self.inserter.function.dfg.make_constant(zero, Type::field())
+                }
+            }
+            Type::Array(element_types, len) => {
+                if let Some(value) = value {
+                    self.inserter.resolve(value)
+                } else {
+                    let mut array = im::Vector::new();
+                    for _ in 0..*len {
+                        for typ in element_types.iter() {
+                            array.push_back(self.attach_slice_dummies(typ, None, false, max_sizes));
+                        }
+                    }
+                    self.inserter.function.dfg.make_array(array, typ.clone())
+                }
+            }
+            Type::Slice(element_types) => {
+                let (current_size, max_sizes) =
+                    max_sizes.split_first().expect("ICE: Missing internal slice max size");
+                let mut max_size = *current_size;
+                if let Some(value) = value {
+                    let mut slice = im::Vector::new();
+                    match &self.inserter.function.dfg[value].clone() {
+                        Value::Array { array, .. } => {
+                            if is_parent_slice {
+                                max_size = array.len() / element_types.len();
+                            }
+                            for i in 0..max_size {
+                                for (element_index, element_type) in
+                                    element_types.iter().enumerate()
+                                {
+                                    let index_usize = i * element_types.len() + element_index;
+                                    if index_usize < array.len() {
+                                        slice.push_back(self.attach_slice_dummies(
+                                            element_type,
+                                            Some(array[index_usize]),
```

Comment:
can we reduce some of this nesting?

Reply:
I can't do something like this:
```
let array = match &self.inserter.function.dfg[value].clone() {
                        Value::Array { array, .. } => {
                            array
                        }
                        _ => {
                            panic!("Expected an array value");
                        }
                    };
```
as `array` is a temporary value that is dropped at the end of the match statement. Unless I want to sacrifice cloning each array it looks like nesting needs to stay.

Reply:
We can do something like

```rust
for (element_index, element_type) in
    element_types.iter().enumerate()
{
    let index_usize = i * element_types.len() + element_index;
    let valid_index = index_usize < array.len();
    let maybe_value = valid_index.then_some(array[index_usize]);
    slice.push_back(self.attach_slice_dummies(
        element_type,
        maybe_value,
        false,
        max_sizes,
    ));
}
```

which removes one level from the inner loop.

Reply:
```rust
                    let mut slice = im::Vector::new();

                    let array = match self.inserter.function.dfg[value].clone() {
                        Value::Array { array, .. } => array,
                        _ => panic!("Expected an array value"),
                    };

                    if is_parent_slice {
                        max_size = array.len() / element_types.len();
                    }
                    for i in 0..max_size {
                        for (element_index, element_type) in element_types.iter().enumerate() {
                            let index_usize = i * element_types.len() + element_index;
                            let valid_index = index_usize < array.len();
                            let maybe_value = valid_index.then_some(array[index_usize]);
                            slice.push_back(self.attach_slice_dummies(
                                element_type,
                                maybe_value,
                                false,
                                max_sizes,
                            ));
                        }
                    }

                    self.inserter.function.dfg.make_array(slice, typ.clone())
```

This compiles for me.

Reply:
Ah I just needed to not borrow `self.inserter.function.dfg[value]`. And creating the maybe value is much cleaner. I will switch, thanks @TomAFrench

Reply:
Hmm we will still error out on `array[index_usize]` though with this snippet:
```
let maybe_value = valid_index.then_some(array[index_usize]);
```
I will look at how we can remove the if statement before pushing though

Reply:
Ok pushed a new version, but I had to keep an if statement around the `maybe_value`

Reply:
> as array is a temporary value that is dropped at the end of the match statement. Unless I want to sacrifice cloning each array it looks like the nesting needs to stay.

You can do that actually, since `im::Vector`'s Clone is `O(1)`

Reply:
> You can do that actually, since `im::Vector`'s Clone is `O(1)`

I went with Tom's suggestion as simply not borrowing also worked but noted for the future.